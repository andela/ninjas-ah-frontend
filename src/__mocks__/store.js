import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import initialState from '../store/initialState';
import apiMiddleware from '../middlewares/apiMiddleware';

export { initialState };
export const middlewares = [thunk, apiMiddleware];
export const mockStore = configureMockStore(middlewares);

export default mockStore(initialState);
