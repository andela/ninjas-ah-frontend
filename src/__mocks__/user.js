export default (Object.user = {
  firstName: 'John',
  lastName: 'Smith',
  username: 'josmi',
  email: 'josmi@email.com',
  image: 'image.jpg'
});

export const userToRegister = {
  ...Object.user,
  password: 'Abcd1234!',
  confirmPassword: 'Abcd1234!'
};

export const mismatchedUserPassword = {
  ...Object.user,
  password: '12345678',
  confirmPassword: '1234abcd'
};

export const wrongUserPassword = {
  ...Object.user,
  password: 'Abcd1234'
};

export const resetPassword = {
  password: '12345678',
  confirmPassword: '12345678'
};

export const mismatchedResetPassword = {
  password: '12345678',
  confirmPassword: '1234abcd'
};

export const matchedResetPassword = {
  passwordOne: 'Brazzaville10!',
  passwordTwo: 'Brazzaville10!'
};

export const sendEmail = { email: 'noreplay@gmail.com' };
export const fakeEmail = { email: 'fsdfsdfsdfdsf--gmail.com' };
