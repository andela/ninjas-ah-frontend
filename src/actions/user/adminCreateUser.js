import { userActionsTypes } from '../../actions-types';
import { apiAction } from '../../helpers';

export default formData => dispatch => dispatch(apiAction({
  method: 'post',
  url: '/users',
  data: { ...formData },
  onStart: userActionsTypes.ADMIN_CREATE_USER_START,
  onEnd: userActionsTypes.ADMIN_CREATE_USER_END,
  onSuccess: userActionsTypes.ADMIN_CREATE_USER_SUCCESS,
  onFailure: userActionsTypes.ADMIN_CREATE_USER_FAILURE
}));
