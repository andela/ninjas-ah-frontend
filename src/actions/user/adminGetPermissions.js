import { userActionsTypes } from '../../actions-types';
import { apiAction } from '../../helpers';

export default userType => dispatch => dispatch(apiAction({
  method: 'get',
  url: `/permissions/${userType || ''}`,
  onStart: userActionsTypes.ADMIN_GET_PERMISSIONS_START,
  onEnd: userActionsTypes.ADMIN_GET_PERMISSIONS_END,
  onSuccess: userActionsTypes.ADMIN_GET_PERMISSIONS_SUCCESS,
  onFailure: userActionsTypes.ADMIN_GET_PERMISSIONS_FAILURE
}));
