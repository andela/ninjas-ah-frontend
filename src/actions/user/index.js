import * as getUser from './getUser';
import signup from './signup';
import socialAuth from './socialAuth';

export { getUser, signup, socialAuth };
