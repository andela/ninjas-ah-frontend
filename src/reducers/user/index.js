import { user as initialState } from '../../store/initialState';
import getUserReducer from './getUserReducer';
import signupReducer from './signupReducer';
import loginReducer from './loginReducer';
import editProfileReducer from './editProfileReducer';
import uploadImageReducer from './uploadImageReducer';
import forgotPasswordReducer from './forgotPasswordReducer';
import updatePasswordReducer from './updatePasswordReducer';

export default (state = initialState, action) => {
  const getUser = getUserReducer(state, action);
  const signup = signupReducer(state, action);
  const login = loginReducer(state, action);
  const editProfile = editProfileReducer(state, action);
  const uploadImage = uploadImageReducer(state, action);
  const forgotPassword = forgotPasswordReducer(state, action);
  const updatePassword = updatePasswordReducer(state, action);
  return (
    getUser
    || signup
    || editProfile
    || uploadImage
    || login
    || forgotPassword
    || updatePassword
    || state
  );
};
