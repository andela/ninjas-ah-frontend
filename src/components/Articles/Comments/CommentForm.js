import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import TextareaAutosize from 'react-autosize-textarea';
// import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
// import {
//   faThumbsUp,
//   faClock,
//   faPenAlt,
//   faTrash,
//   faCommentAlt
// } from '@fortawesome/free-solid-svg-icons';
import './Comments.scss';
import { createComment } from '../../../actions';

export class CommentForm extends Component {
  state = { loading: true, comment: '' };

  componentWillMount() {
    const { comments } = this.props;
    // createComment(this.props.slug);
    this.setState({ comments });
  }

  componentWillReceiveProps(nextProps) {
    const { errors, comments, message } = nextProps;
    console.log('will', message, comments);
    if (errors) {
      this.setState({ errors });
    }
    this.setState({ loaded: true });
  }

  onChange = (e) => {
    this.setState({ [e.target.name]: e.target.value });
  };

  onSubmit = async (e) => {
    e.preventDefault();
    const { isAuth, slug } = this.props;
    if (isAuth) {
      await this.props.createComment({ comment: this.state.comment, slug });
    } else {
      this.setState({ errors: { token: 'Failed to authenticate token' } });
    }
  };

  render() {
    const { isAuth, profile, slug, loading } = this.props;
    const { errors, comment, message } = this.state;
    console.log('loading', loading, message, errors && errors);
    return (
      <div id="comment-form-wrapper">
        {!isAuth || (errors && errors.token) ? (
          <div className="border b-light medium-padding">
            To comment,{' '}
            <Link
              className="bold text-info"
              to={`/login?redirect=articles/${slug}#comment-form-wrapper`}
            >
              Login
            </Link>{' '}
            first
          </div>
        ) : (
          <div className="border b-light medium-padding">
            You are commenting as{' '}
            <Link to="/" className="bold text-success">
              {profile.firstName} {profile.lastName}
            </Link>
          </div>
        )}
        <form onSubmit={this.onSubmit}>
          <div className="input-field">
            <TextareaAutosize
              name="comment"
              onChange={this.onChange}
              value={comment}
              placeholder="Write your comment here..."
              style={{ minHeight: 12 }}
              id="commentBody"
              className="shadow-1"
            />
          </div>
          <div className="input-field">
            <button type="submit" className="button success bold text-white radius-5">
              Comment
            </button>
            {loading ? <span className="medium-padding radius-5">loading</span> : ''}
            {errors && errors.token ? (
              <span className="medium-padding radius-5 text-danger">You need to login first</span>
            ) : (
              ''
            )}
            {errors && errors.message ? (
              <span className="medium-padding radius-5 text-danger">{errors.message}</span>
            ) : (
              ''
            )}
          </div>
        </form>
        <div className="clear" />
      </div>
    );
  }
}

CommentForm.propTypes = {
  createComment: PropTypes.func.isRequired,
  isAuth: PropTypes.bool,
  profile: PropTypes.object,
  comments: PropTypes.array,
  slug: PropTypes.string,
  comment: PropTypes.string,
  rating: PropTypes.number,
  params: PropTypes.object,
  match: PropTypes.object,
  errors: PropTypes.object,
  loading: PropTypes.bool,
  message: PropTypes.object
};

const mapStateToProps = ({
  user: { isAuth, profile },
  comments: { createComment: { comments, message, errors, loading } }
}) => ({ isAuth, profile, comments, message, errors, loading });

export default connect(
  mapStateToProps,
  { createComment }
)(CommentForm);
