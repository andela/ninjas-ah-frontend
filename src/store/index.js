import 'dotenv/config';
import { createStore, applyMiddleware, combineReducers } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension/developmentOnly';
import thunk from 'redux-thunk';
import initialState from './initialState';
import reducers from '../reducers';
import apiMiddleware from '../middlewares/apiMiddleware';

const middlewares = [thunk, apiMiddleware];

export default createStore(
  combineReducers(reducers),
  initialState,
  composeWithDevTools(applyMiddleware(...middlewares))
);
