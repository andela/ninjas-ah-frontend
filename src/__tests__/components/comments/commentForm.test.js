import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import store from '../../../__mocks__/store';
import CommentForm, { CommentForm as CommentFormComponent } from '../../../components/Articles/Comments/CommentForm';
import { shallow, mount } from '../../../../config/enzymeConfig';
import user from '../../../__mocks__/user';

const props = {
  errors: { error: ['12'] },
  comments: [
    {
      body: 'Hello John Doe',
      userId: 1,
      commentAuthor: {
        firstName: 'John',
        lastName: 'Doe',
        username: 'johndoe',
        image: 'image.png'
      },
      createdAt: '2019-01-01'
    }
  ],

  createComment: jest.fn(),
  isAuth: true,
  profile: {
    id: 1,
    ...user
  }
};
const state = { loading: true, comment: 'this is the comment' };

describe('<CommentComponent />', () => {
  const component = shallow(<CommentFormComponent {...props} />);
  test('should render a <CommentFormComponent /> component ', () => {
    expect(component).toMatchSnapshot();
  });
  test('should set state', () => {
    component.setState({ ...state });
    expect(component).toHaveLength(1);
  });

  test('should call onChange method when the comment is provided', () => {
    const component = shallow(<CommentFormComponent {...props} />);
    const spy = jest.spyOn(component.instance(), 'onChange');
    component.instance().forceUpdate();
    component.setProps({ ...props, errors: { errors: ['error1'] } });
    const event = { target: { value: 'body' } };
    const textArea = component.find('#commentBody');
    textArea.simulate('change', event);
    expect(textArea.length).toBe(1);
    expect(spy).toHaveBeenCalled();
  });

  test('should submit comment', () => {
    const component = shallow(<CommentFormComponent state={state} {...props} />);
    const instance = component.instance();

    const fakeEvent = { preventDefault: jest.fn() };
    instance.onSubmit(fakeEvent);
  });

  test('', () => {
    const component = shallow(<CommentFormComponent state={state} {...{ ...props, isAuth: false }} />);

    component.setState({ errors: { token: 'some' } });
    const fakeEvent = { preventDefault: () => {} };
  });

  test('render the props', () => {
    const component = mount(<Provider store={store}>
        <MemoryRouter>
          <CommentForm {...props} />
        </MemoryRouter>
      </Provider>);

    expect(component).toHaveLength(1);
  });
});
