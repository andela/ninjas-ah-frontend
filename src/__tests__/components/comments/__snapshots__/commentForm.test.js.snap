// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CommentComponent /> should render a <CommentFormComponent /> component  1`] = `ShallowWrapper {}`;
